Add GATK contamination check to complement VerifyBamID2

dorotejavujinovic opened the pull request:
<!--
# nf-core/raredisease pull request

Many thanks for contributing to nf-core/raredisease!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/raredisease/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [x] This comment contains a description of changes (with reason)
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If necessary, also make a PR on the nf-core/raredisease _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repo
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Make sure your code lints (`nf-core lint .`).
- [ ] Documentation in `docs` is updated
- [x] `CHANGELOG.md` is updated
- [ ] `README.md` is updated

## Description

Adds GATK-based contamination detection to complement VerifyBamID2.

### Background

- **VerifyBamID2** works well for WGS but has significant limitations with WES data
- **GATK CalculateContamination** performs better on targeted sequencing (WES)  
- Having both methods for WGS provides cross-validation

### Implementation

- New subworkflow: `CONTAMINATION_CHECK` using GATK4 GetPileupSummaries and CalculateContamination
- New module: `PARSE_CONTAMINATION` for MultiQC integration
- Conditional interval handling: WGS (genome-wide) vs WES (target regions)
- MultiQC configuration with color-coded thresholds

### Usage
```groovy
params.run_contamination = true
params.contamination_sites = "small_exac_common_3.hg38.vcf.gz"
params.contamination_sites_tbi = "small_exac_common_3.hg38.vcf.gz.tbi"
```

### Testing

Tested on both WGS and WES samples with successful integration into MultiQC reports.

Commit:
Add GATK contamination check for WES/WGS samples

- Add CONTAMINATION_CHECK subworkflow using GATK4
- Add PARSE_CONTAMINATION module for MultiQC integration
- Add GATK4 GetPileupSummaries and CalculateContamination modules
- Implement conditional intervals handling (WGS vs WES)
- Update workflow to integrate contamination check after QC_BAM
- Configure MultiQC to display contamination results

dorotejavujinovic committed:
Remove deepvariant_parabricks config inclusion

Commit:
Fix trailing whitespace and EOF newlines

dorotejavujinovic committed:
Refactor GATK contamination settings in YAML

Updated GATK contamination configuration for clarity and consistency.

dorotejavujinovic committed:
Change default test branch from 'dev' to 'master'

dorotejavujinovic committed:
Change default branch from 'master' to 'dev'

dorotejavujinovic committed:
Update CHANGELOG with GATK contamination features

Introduced GATK contamination check for WES/WGS samples, added new parameters and subworkflow, and updated MultiQC configuration.

dorotejavujinovic committed:
Refactor workflow includes and MultiQC file handling

dorotejavujinovic committed:
Merge branch 'dev' into gatk-contamination-clean

ramprasadn commented on conf/modules/align_bwa_bwamem2_bwameme.config:
```diff
@@ -58,7 +58,7 @@ process {
         ext.prefix = { "${meta.id}_sorted_md" }
         publishDir = [
-            enabled: !params.save_mapped_as_cram,
+            enabled: true,
```
I am curious why do you want to change this?

ramprasadn commented on conf/modules/align_bwa_bwamem2_bwameme.config:
```diff
@@ -67,7 +67,7 @@ process {
     withName: '.*ALIGN:ALIGN_BWA_BWAMEM2_BWAMEME:SAMTOOLS_INDEX_MARKDUP' {
         publishDir = [
-            enabled: !params.save_mapped_as_cram,
+            enabled: true,
```
and this

ramprasadn commented on modules/local/parse_contamination/main.nf:
```diff
@@ -0,0 +1,61 @@
+    script:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    #!/usr/bin/env python3
```
Can you make this a module binary? We have had issues in the past with some systems interpreting indents differently.

ramprasadn commented on modules/local/parse_contamination/main.nf:
```diff
@@ -0,0 +1,61 @@
+        v.write('"${task.process}":\\n')
+        v.write('    python: "3.11"\\n')
+    """
+}
```
Also can you add a stub section?

ramprasadn commented on CHANGELOG.md:
```diff
@@ -3,6 +3,19 @@
+- New parameters: `run_contamination`, `contamination_sites`, `contamination_sites_tbi`
+- CONTAMINATION_CHECK subworkflow using GATK4 GetPileupSummaries and CalculateContamination
+- PARSE_CONTAMINATION module for MultiQC integration
+- Contamination results displayed in MultiQC with color-coded thresholds
```
You can add your log entries to 2.7.0dev since its the one in development. And don't forget to link the PR to your entries ;)  

Also, we have a separate table for parameters and new tools under the `##Fixed` section of 2.7.0dev, so you can add that information there.

ramprasadn commented on subworkflows/local/contamination_check/main.nf:
Could you add a test for this subworkflow? We are currently in the process of adding subworkflow level tests using nf-test, so it would be fantastic if you can include one for this subworkflow.

ramprasadn reviewed (commented):
Thanks for the PR @dorotejavujinovic!

Commit:
Address PR review: refactor parse_contamination, fix changelog, add nf-tests

- Convert parse_contamination inline Python to module binary (bin/)
- Add stub section to parse_contamination module
- Move CHANGELOG entries into 2.7.0dev section with PR links and tables
- Revert unrelated publishDir changes in align config
- Add nf-test subworkflow tests for contamination_check (WGS + WES)

Commit:
Fix missing EOF newline and extra blank line in CHANGELOG

Commit:
Fix malformed modules.json entry for gatk4/getpileupsummaries

Missing closing brace and installed_by field caused nf-core lint to
crash with TypeError during dependency recreation.

Commit:
Fix modules.json SHA for gatk4 contamination modules

Update git_sha to ae8cd884f895 which matches the actual installed
module versions (GATK 4.6.2.0 with nf-core code style updates).

Commit:
Fix trailing whitespace in raredisease.nf

ramprasadn committed:
Merge branch 'dev' of github.com:nf-core/raredisease into pr-758

dorotejavujinovic committed:
Address review: topic-channel versions, skip_tools gating, move params to main.nf, sync GATK modules, fix svdb/merge newline

dorotejavujinovic committed:
Merge remote-tracking branch 'upstream/dev' into gatk-contamination-clean

# Conflicts:
#	CHANGELOG.md

ramprasadn committed:
Merge branch 'dev' into gatk-contamination-clean

ramprasadn commented on workflows/raredisease.nf:
```diff
@@ -405,6 +411,35 @@ workflow RAREDISEASE {
         val_target_bed,
         skip_ngsbits
     )
+    ch_versions = ch_versions.mix(QC_BAM.out.versions)
```
Could you remove this line @dorotejavujinovic ? QC bam does not emit versions anymore. Some of the tests are failing because of this missing `versions` property.

ramprasadn commented on conf/modules/contamination.config:
```diff
@@ -0,0 +1,44 @@
+    //
+    withName: '.*:RAREDISEASE:PARSE_CONTAMINATION' {
+        ext.prefix = { "${meta.id}_contamination" }
+        publishDir = [
```
Sorry I forgot to mention this in my previous review. But we have moved away from using publishDir in the pipeline in favour of workflow outputs. https://docs.seqera.io/nextflow/tutorials/workflow-outputs

dorotejavujinovic committed:
Address review: drop stale QC_BAM.out.versions mix; publish contamination outputs via workflow outputs instead of publishDir

ramprasadn commented on conf/modules/contamination_check.config:
Could you change the name of this config file from `contamination.config` to `contamination_check.config`?

ramprasadn commented on bin/parse_contamination.py:
I'd avoid module binaries since they add a dependency on Wave, which could end up causing issues. You can just move the scripts from the local modules into the root bin folder instead. That should work the same.

ramprasadn commented on workflows/raredisease.nf:
```diff
@@ -27,13 +27,15 @@ include { SPRING_DECOMPRESS as SPRING_DECOMPRESS_TO_R1_FQ   } from '../modules/n
 include { RENAME_ALIGN_FILES as RENAME_BAM } from '../modules/local/rename_align_files'
 include { RENAME_ALIGN_FILES as RENAME_BAI } from '../modules/local/rename_align_files'
+include { CREATE_HGNCIDS_FILE              } from '../modules/local/create_hgncids_file'
+include { CREATE_PEDIGREE_FILE             } from '../modules/local/create_pedigree_file'
```
```suggestion
```

ramprasadn commented on nextflow_schema.json:
```diff
@@ -652,8 +669,8 @@
-                    "help_text": "Multiple tools can be specified, separated by commas.",
-                    "pattern": "^((fastp|gens|germlinecnvcaller|peddy|smncopynumbercaller|vcf2cytosure|fastqc|ngsbits|mitosalt)?,?)*(?<!,)$"
+                    "help_text": "Multiple tools can be specified, separated by commas. The GATK contamination check additionally requires `contamination_sites` to be set; it is skipped automatically when that file is not provided.",
+                    "pattern": "^((fastp|gens|germlinecnvcaller|peddy|smncopynumbercaller|vcf2cytosure|fastqc|ngsbits|mitosalt|contamination)?,?)*(?<!,)$"
```
I would suggest that we change this parameter from `contamination` to `gatkcontamination`. There is verifybamid in the pipeline which also detects contamination, and we are planning to add somalier contamination as well. It would be nice to keep it specific.

ramprasadn reviewed (commented):
Just a few minor comments from me @dorotejavujinovic. Otherwise looks good 👍🏻

dorotejavujinovic committed:
Address review: rename contamination_check.config; move parse_contamination.py to root bin; remove duplicate CREATE_* includes; rename skip_tools option to gatkcontamination

dorotejavujinovic committed:
Merge remote-tracking branch 'upstream/dev' into gatk-contamination-clean

# Conflicts:
#	main.nf
